ethash/src/progpow.rs

```diff
@@ -0,0 +1,636 @@
+		self.z = 36969u32.wrapping_mul(self.z & 65535).wrapping_add(self.z >> 16);
+		self.w = 18000u32.wrapping_mul(self.w & 65535).wrapping_add(self.w >> 16);
+		let mwc = (self.z << 16).wrapping_add(self.w);
+		self.jsr ^= self.jsr << 17;
+		self.jsr ^= self.jsr >> 13;
+		self.jsr ^= self.jsr << 5;
+		self.jcong = 69069u32.wrapping_mul(self.jcong).wrapping_add(1234567);
+
+		(mwc ^ self.jcong).wrapping_add(self.jsr)
+	}
+}
+
+fn fill_mix(seed: u64, lane_id: u32) -> [u32; PROGPOW_REGS] {
+	// Use FNV to expand the per-warp seed to per-lane
+	// Use KISS to expand the per-lane seed to fill mix
+	let z = fnv1a_hash(FNV_HASH, seed as u32);
+	let w = fnv1a_hash(z, (seed >> 32) as u32);
+	let jsr = fnv1a_hash(w, lane_id);
+	let jcong = fnv1a_hash(jsr, lane_id);
+
+	let mut rnd = Kiss99::new(z, w, jsr, jcong);
+
+	let mut mix = [0; PROGPOW_REGS];
+
+	debug_assert_eq!(PROGPOW_REGS, 32);
+	unroll! {
+		for i in 0..32 {
+			mix[i] = rnd.next_u32();
+		}
+	}
+
+	mix
+}
+
+// Merge new data from b into the value in a. Assuming A has high entropy only
+// do ops that retain entropy even if B is low entropy (IE don't do A&B)
+fn merge(a: u32, b: u32, r: u32) -> u32 {
+	match r % 4 {
+		0 => a.wrapping_mul(33).wrapping_add(b),
+		1 => (a ^ b).wrapping_mul(33),
+		2 => a.rotate_left(((r >> 16) % 31) + 1) ^ b,
+		_ => a.rotate_right(((r >> 16) % 31) + 1) ^ b,
+	}
+}
+
+fn math(a: u32, b: u32, r: u32) -> u32 {
+	match r % 11 {
+		0 => a.wrapping_add(b),
+		1 => a.wrapping_mul(b),
+		2 => ((a as u64).wrapping_mul(b as u64) >> 32) as u32,
+		3 => a.min(b),
+		4 => a.rotate_left(b),
+		5 => a.rotate_right(b),
+		6 => a & b,
+		7 => a | b,
+		8 => a ^ b,
+		9 => a.leading_zeros() + b.leading_zeros(),
+		_ => a.count_ones() + b.count_ones(),
+	}
+}
+
+fn progpow_init(seed: u64) -> (Kiss99, [u32; PROGPOW_REGS], [u32; PROGPOW_REGS]) {
+	let z = fnv1a_hash(FNV_HASH, seed as u32);
+	let w = fnv1a_hash(z, (seed >> 32) as u32);
+	let jsr = fnv1a_hash(w, seed as u32);
+	let jcong = fnv1a_hash(jsr, (seed >> 32) as u32);
+
+	let mut rnd = Kiss99::new(z, w, jsr, jcong);
+
+	// Create a random sequence of mix destinations for merge() and mix sources
+	// for cache reads guarantees every destination merged once and guarantees
+	// no duplicate cache reads, which could be optimized away. Uses
+	// Fisher-Yates shuffle.
+	let mut mix_seq_dst = [0u32; PROGPOW_REGS];
+	let mut mix_seq_cache = [0u32; PROGPOW_REGS];
+	for i in 0..mix_seq_dst.len() {
+		mix_seq_dst[i] = i as u32;
+		mix_seq_cache[i] = i as u32;
+	}
+
+	for i in (1..mix_seq_dst.len()).rev() {
+		unsafe {
+			// NOTE: `i` takes values from the range [1..PROGPOW_REGS] and `j`
+			// takes values from the the range [0..i]. This way it is guaranteed
+			// that the indices are always within the range of `mix_seq_dst` and
+			// `mix_seq_cache` and we can skip the bounds checking.
+			let j = rnd.next_u32() as usize % (i + 1);
+			::std::ptr::swap(&mut mix_seq_dst[i], mix_seq_dst.get_unchecked_mut(j));
```

Comment:
Does swap improve performance over just u32 assigns (I hoped the compiler would optimize that)?

Reply:
`unchecked_mut` was only used on `j`, since the compiler can't prove that it will always be in-bounds, so it must add a bounds check. You can check this by looking at the generated assembly, at least it generated different code when I wrote this. Anyway, even with the bounds check it probably doesn't make a difference performance-wise because of branch prediction, I couldn't reliably microbenchmark this on my laptop so I'm removing all unsafe code.

Reply:
Right, I verified it with this [godbolt example](https://rust.godbolt.org/z/JE5GZU) so you are right cool!